templates/consul_systemd.service.j2

```
@@ -23,6 +23,7 @@ Environment=CONSUL_UI_LEGACY=true
 {% endif %}
 ExecStartPre=-/bin/mkdir -m 0750 -p {{ consul_run_path }}
 ExecStartPre=/bin/chown -R {{ consul_user }}:{{ consul_group }} {{ consul_run_path }}
+ExecStartPre={{ consul_binary }} validate -quiet {{ consul_config_path }}/config.json {{ consul_configd_path }}
```

You may want to use the same syntax as bellow 

```
{{ consul_bin_path }}/consul
```

---

Two questions:

1. Why do we need to validate on service startup? We already do it when configuring the task.
2. Why only do this for systemd? There are several more service manager files in `templates` (namely, for SysV init, macOS launchctl, BSD, etc. This change should be applied to all of them, or not at all.

I would suggest dropping this change. I think it's enough to validate the config file after configuring it in `tasks/config.yml`.

---

@abarbare I made it consistent using `consul_binary` on the other call, so the binary name isn't hardcoded anywhere.

@nre-ableton 
1.  It ensure any external configuration change (manual change for exemple) will be validated. The validate gives you a non 0 return code and system will show you the error right away, instead of failing silently (a systemctl restart consul won't tell you anything about any issue). I thinks it's a nice and free addition, but not mandatory. If you really prefer to remove it I'll do it. The validation during the playbook run is the most important part.  
2. I only have systemd based system to test it, but I can probably use molecule to validated it ? (never done any automated testing with ansible). 

If you are willing to keep the change I can look into adding and testing the change into other service manager init script.

---

> @abarbare I made it consistent using `consul_binary` on the other call, so the binary name isn't hardcoded anywhere.

BTW, I didn't mention this in my last review (since I advocated for removal of that change), but I appreciated use of `consul_binary`. The _other_ code should be changed and we shouldn't use `{{ consul_bin_path }}/consul` anymore, but that's for a future cleanup PR.

> 1. It ensure any external configuration change (manual change for exemple) will be validated. The validate gives you a non 0 return code and system will show you the error right away, instead of failing silently (a systemctl restart consul won't tell you anything about any issue). I thinks it's a nice and free addition, but not mandatory. If you really prefer to remove it I'll do it. The validation during the playbook run is the most important part.
> 
> 2. I only have systemd based system to test it, but I can probably use molecule to validated it ? (never done any automated testing with ansible).
>
> If you are willing to keep the change I can look into adding and testing the change into other service manager init script.

@bastienwirtz I agree that the verification check would be useful to have before starting the service, and you've made a good case for having it here. However, we should try to maintain feature parity and consistency in this role and not get in the habit of adding features for some systems but not others. I did a brief bit of digging, and fully supporting it on all systems would involve the following:

- FreeBSD (`consul_bsdinit.j2`): just a matter of adding the extra command in the `consul_start()` function
- SysV init (`consul_sysvinit.j2`): same thing, except the function is named `start()`
- Debian non-systemd (`consul_debianinit.j2`): we could actually remove this, because systemd is the [default service manager on Debian since version 8 (jessie)](https://wiki.debian.org/systemd), which [isn't supported by this role anymore](https://github.com/ansible-community/ansible-consul/pull/549).
- macOS (`consul_launchctl.plist.j2`): this is where things start to get tricky. AFAIK, launchctl doesn't support anything like `ExecStartPre` in systemd. It *might* be possible to make a secondary service and check the result of that, but I'm not sure that this would work if it were a one-shot service. I don't think that launchctl allows you to do things like `&&` in the `ProgramArguments` array.
- Windows (`tasks/windows.yml`): on Windows, we don't need a template, because [`win_service`](https://docs.ansible.com/ansible/latest/collections/ansible/windows/win_service_module.html) does everything we need. The implementation here might be similar to macOS by using a separate service. I'm also not sure whether `win_service` allows you to have compounded commands joined by `&&`.

I understand that the burden of implementing this on all platforms (especially Mac/Windows) is unfair to ask of you. I would therefore suggest that we compromise with the following approach:

1. Add a separate commit to remove support for Debian non-systemd. This should just be a matter of removing the `consul_debianinit.j2` file, and the corresponding `template` task in `tasks/nix.yml`. We could also even unofficially support it by changing the first condition of the "Create SYSV init script" task to be `ansible_service_mgr == "sysvinit"` instead of the current `ansible_service_mgr != "systemd"` (which is arguably more correct anyway).
2. Add support for FreeBSD and SysV init, as detailed above.
3. We can make an issue in the backlog to provide the feature for Mac/Windows systems. This way, the lack of feature parity is at least documented, and perhaps someone from the community will have a good solution for either (or both) of these systems.

Does that sound like a good approach? Alternately, we can drop support for this in the service start if you think it's too much of a pain.

---

Sounds like a plan @nre-ableton! Thanks for digging into it, I'll work on it, glad to contribute :)

---

Updated @nre-ableton :+1: